File: backend/src/main/java/com/technicalescaperoom/backend/controller/player/PlayerHintController.java
package com.technicalescaperoom.backend.controller.player;

import com.technicalescaperoom.backend.config.security.PlayerPrincipal;
import com.technicalescaperoom.backend.dto.player.PlayerHintsResponseDto;
import com.technicalescaperoom.backend.service.HintService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/player/game")
@RequiredArgsConstructor
public class PlayerHintController {

    private final HintService hintService;

    @GetMapping("/hints")
    public ResponseEntity<PlayerHintsResponseDto> getPlayerHints(@AuthenticationPrincipal PlayerPrincipal principal) {
        PlayerHintsResponseDto response = hintService.getHintsForPlayer(principal);
        return ResponseEntity.ok(response);
    }

    @PostMapping("/hints/{levelNumber}/{stageNumber}/{hintNumber}")
    public ResponseEntity<com.technicalescaperoom.backend.dto.player.HintUseResponseDto> useHint(
            @AuthenticationPrincipal PlayerPrincipal principal,
            @PathVariable Integer levelNumber,
            @PathVariable Integer stageNumber,
            @PathVariable Integer hintNumber) {
        return ResponseEntity.ok(hintService.useHint(principal, levelNumber, stageNumber, hintNumber));
    }
}

File: backend/controllers/player/playerHintController.js
const hintService = require('../../services/hintService');

class PlayerHintController {
  async getPlayerHints(req, res, next) {
    try {
      const response = await hintService.getHintsForPlayer(req.player);
      res.json(response);
    } catch (err) {
      next(err);
    }
  }

  async useHint(req, res, next) {
    try {
      const levelNumber = parseInt(req.params.levelNumber, 10);
      const stageNumber = parseInt(req.params.stageNumber, 10);
      const hintNumber = parseInt(req.params.hintNumber, 10);
      const response = await hintService.useHint(req.player, levelNumber, stageNumber, hintNumber);
      res.json(response);
    } catch (err) {
      next(err);
    }
  }
}

module.exports = new PlayerHintController();
